src/os_posix/os_fs.c

```diff
@@ -501,10 +560,26 @@ __posix_file_truncate(WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session, wt_of
     session = (WT_SESSION_IMPL *)wt_session;
     pfh = (WT_FILE_HANDLE_POSIX *)file_handle;
 
+    __wt_verbose(session, WT_VERB_FILEOPS, "%s, file-truncate: size=%" PRId64
+                                           ","
+                                           "mapped size=%" WT_SIZET_FMT "\n",
+      file_handle->name, len, pfh->mmap_size);
+
+    if ((size_t)len != pfh->mmap_size)
+        __wt_prepare_remap_resize_file(file_handle, wt_session);
+
     WT_SYSCALL_RETRY(ftruncate(pfh->fd, len), ret);
-    if (ret == 0)
-        return (0);
-    WT_RET_MSG(session, ret, "%s: handle-truncate: ftruncate", file_handle->name);
+    if (ret != 0) {
+        __wt_err(session, ret, "%s: handle-truncate: ftruncate", file_handle->name);
+        if ((size_t)len != pfh->mmap_size)
+            __wt_release_without_remap(file_handle);
+        return (ret);
+    }
```

Comment:
Staring at this, I think there's a race.

Imagine a thread calls truncate and `len != pfh->mmapsize`, so it calls `__wt_prepare_remap_resize_file()`. While it's waiting to acquire the ` mmap_resizing` lock, another thread resizes the mmap region such that `len == pfh->mmapsize`, then the thread returns, holding the lock. The truncate code then checks that `len != pf->mmapsize`, and because they are equal, it doesn't ever release the lock.

I suspect what has to happen here is we don't test for the length, once we've acquired the lock, we have to release it regardless.

The fallocate code appears the same.

There's also a possibility with `pfh->mmap_file_mappable`, I think. If that gets turned off while we're acquiring the lock, we won't release the lock, again, we have to always release the lock.

Something like this, maybe:
```
 diff --git a/src/os_posix/os_fs.c b/src/os_posix/os_fs.c
index 1de3f9162..35ee10c76 100644
--- a/src/os_posix/os_fs.c
+++ b/src/os_posix/os_fs.c
@@ -556,6 +556,7 @@ __posix_file_truncate(WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session, wt_of
     WT_DECL_RET;
     WT_FILE_HANDLE_POSIX *pfh;
     WT_SESSION_IMPL *session;
+    bool remap;
 
     session = (WT_SESSION_IMPL *)wt_session;
     pfh = (WT_FILE_HANDLE_POSIX *)file_handle;
@@ -565,21 +566,20 @@ __posix_file_truncate(WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session, wt_of
                                            "mapped size=%" WT_SIZET_FMT "\n",
       file_handle->name, len, pfh->mmap_size);
 
-    if ((size_t)len != pfh->mmap_size)
+    remap = size_t)len != pfh->mmap_size;
+    if (remap)
         __wt_prepare_remap_resize_file(file_handle, wt_session);
 
     WT_SYSCALL_RETRY(ftruncate(pfh->fd, len), ret);
-    if (ret != 0) {
-        __wt_err(session, ret, "%s: handle-truncate: ftruncate", file_handle->name);
-        if ((size_t)len != pfh->mmap_size)
-            __wt_release_without_remap(file_handle);
-        return (ret);
-    }
-
-    if ((size_t)len != pfh->mmap_size)
-        __wt_remap_resize_file(file_handle, wt_session);
 
-    return (ret);
+    if (remap) {
+       if (ret == 0)
+           __wt_remap_resize_file(file_handle, wt_session);
+       else
+           __wt_release_without_remap(file_handle);
+    }
+    WT_RET_MSG( __wt_err(session, ret, "%s: handle-truncate: ftruncate", file_handle->name);
+    return (0);
 }
 #endif

```

Reply:
Great catch!